### ISSUE TITLE
train_input_images oversaturated / gamma

### ISSUE THREAD

**Issue by turowicz** [tensorflow/models#9226]
The input images the TensorBoard shows me are strangely augmented in terms of colour saturation and gamma. Any idea how can I fix it? Is it the wrong file format or compression?

My augmentation options:

```
train_config {
  batch_size: 32
  data_augmentation_options {
    random_horizontal_flip {
    }
  }
  data_augmentation_options {
    random_scale_crop_and_pad_to_square {
      output_size: 640
      scale_min: 0.10000000149011612
      scale_max: 2.0
    }
  }
  sync_replicas: true
  ...
}
```

![image](https://user-images.githubusercontent.com/1055391/92909461-6befc900-f427-11ea-9301-dc77d171c291.png)

**Comment by turowicz** [comment]
Another example:

![image](https://user-images.githubusercontent.com/1055391/92926988-30f89000-f43d-11ea-9b7c-d62c752d31b6.png)

**Comment by turowicz** [comment]
The color / brightness continue even with all augmentation removed.

**Comment by turowicz** [comment]
Another question: shouldn't I have bounding boxes displayed in there?

**Comment by ravikyram** [comment]
@turowicz 

Please, fill [issue template.](https://github.com/tensorflow/models/issues/new/choose).
Please, let us know which pretrained model you are using and share related code .Thanks!

**Comment by turowicz** [comment]
@ravikyram 

# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am using the latest TensorFlow Model Garden release and TensorFlow 2.
- [x] I am reporting the issue to the correct repository. (Model Garden official or research directory)
- [x] I checked to make sure that this issue has not already been filed.

## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/tree/master/research/object_detection/model_main_tf2.py

## 2. Describe the bug

(See first post in this thread)

## 3. Steps to reproduce

1. Run training.
2. Observer Images section in TensorBoard.
3. Input images have incorrect colours and contrasts.

Same thing happens when you skip step 2.

## 4. Expected behavior

I expect to see input images exactly as they are in my training folder.

## 5. Additional context

Eficientnet D1 from https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md

Config:

```
# SSD with EfficientNet-b1 + BiFPN feature extractor,
# shared box predictor and focal loss (a.k.a EfficientDet-d1).
# See EfficientDet, Tan et al, https://arxiv.org/abs/1911.09070
# See Lin et al, https://arxiv.org/abs/1708.02002
# Trained on COCO, initialized from an EfficientNet-b1 checkpoint.
#
# Train on TPU-8

model {
  ssd {
    num_classes: 7
    image_resizer {
      keep_aspect_ratio_resizer {
        min_dimension: 640
        max_dimension: 640
        pad_to_max_dimension: true
      }
    }
    feature_extractor {
      type: "ssd_efficientnet-b1_bifpn_keras"
      conv_hyperparams {
        regularizer {
          l2_regularizer {
            weight: 3.9999998989515007e-05
          }
        }
        initializer {
          truncated_normal_initializer {
            mean: 0.0
            stddev: 0.029999999329447746
          }
        }
        activation: SWISH
        batch_norm {
          decay: 0.9900000095367432
          scale: true
          epsilon: 0.0010000000474974513
        }
        force_use_bias: true
      }
      bifpn {
        min_level: 3
        max_level: 7
        num_iterations: 4
        num_filters: 88
      }
    }
    box_coder {
      faster_rcnn_box_coder {
        y_scale: 1.0
        x_scale: 1.0
        height_scale: 1.0
        width_scale: 1.0
      }
    }
    matcher {
      argmax_matcher {
        matched_threshold: 0.5
        unmatched_threshold: 0.5
        ignore_thresholds: false
        negatives_lower_than_unmatched: true
        force_match_for_each_row: true
        use_matmul_gather: true
      }
    }
    similarity_calculator {
      iou_similarity {
      }
    }
    box_predictor {
      weight_shared_convolutional_box_predictor {
        conv_hyperparams {
          regularizer {
            l2_regularizer {
              weight: 3.9999998989515007e-05
            }
          }
          initializer {
            random_normal_initializer {
              mean: 0.0
              stddev: 0.009999999776482582
            }
          }
          activation: SWISH
          batch_norm {
            decay: 0.9900000095367432
            scale: true
            epsilon: 0.0010000000474974513
          }
          force_use_bias: true
        }
        depth: 88
        num_layers_before_predictor: 3
        kernel_size: 3
        class_prediction_bias_init: -4.599999904632568
        use_depthwise: true
      }
    }
    anchor_generator {
      multiscale_anchor_generator {
        min_level: 3
        max_level: 7
        anchor_scale: 4.0
        aspect_ratios: 1.0
        aspect_ratios: 2.0
        aspect_ratios: 0.5
        scales_per_octave: 3
      }
    }
    post_processing {
      batch_non_max_suppression {
        score_threshold: 9.99999993922529e-09
        iou_threshold: 0.5
        max_detections_per_class: 100
        max_total_detections: 100
      }
      score_converter: SIGMOID
    }
    normalize_loss_by_num_matches: true
    loss {
      localization_loss {
        weighted_smooth_l1 {
        }
      }
      classification_loss {
        weighted_sigmoid_focal {
          gamma: 1.5
          alpha: 0.25
        }
      }
      classification_weight: 1.0
      localization_weight: 1.0
    }
    encode_background_as_zeros: true
    normalize_loc_loss_by_codesize: true
    inplace_batchnorm_update: true
    freeze_batchnorm: false
    add_background_class: false
  }
}
train_config {
  batch_size: 8
  sync_replicas: true
  optimizer {
    momentum_optimizer {
      learning_rate {
        cosine_decay_learning_rate {
          learning_rate_base: 0.07999999821186066
          total_steps: 300000
          warmup_learning_rate: 0.0010000000474974513
          warmup_steps: 2500
        }
      }
      momentum_optimizer_value: 0.8999999761581421
    }
    use_moving_average: false
  }
  fine_tune_checkpoint: "pre-trained-model/checkpoint/ckpt-0"
  num_steps: 300000
  startup_delay_steps: 0.0
  replicas_to_aggregate: 8
  max_number_of_boxes: 100
  unpad_groundtruth_tensors: false
  fine_tune_checkpoint_type: "detection"
  use_bfloat16: true
  fine_tune_checkpoint_version: V2
}
train_input_reader: {
  label_map_path: "annotations/label_map.pbtxt"
  tf_record_input_reader {
    input_path: "annotations/train.record"
  }
}

eval_config: {
  metrics_set: "coco_detection_metrics"
  use_moving_averages: false
  batch_size: 1;
}

eval_input_reader: {
  label_map_path: "annotations/label_map.pbtxt"
  shuffle: false
  num_epochs: 1
  tf_record_input_reader {
    input_path: "annotations/test.record"
  }
}
```

## 6. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.5 (tensorflow:2.3.0-gpu docker img)
- Mobile device name if the issue happens on a mobile device:
- TensorFlow installed from (source or binary): 2.3.0 from docker
- TensorFlow version (use command below): v2.3.0-rc2-23-gb36436b087 2.3.0
- Python version: Python 3.6.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: from docker img
- GPU model and memory: from docker img

**Comment by carlos-vl** [comment]
I have the exact same issue. Anyone could figure this out?

**Comment by carlos-vl** [comment]
I think I found the reason for this behaviour. The input data is encoded as 8bit jpeg or png in the tfrecord files (at least in my case). The images are scaled during the preprocessing step of the feature extractors. Each feature extractor has a different scaling apparently. For example:

https://github.com/tensorflow/models/blob/52b16a1a7f6781e4f65f763e6cf35ca8c4382a4b/research/object_detection/models/ssd_efficientnet_bifpn_feature_extractor.py#L202

https://github.com/tensorflow/models/blob/52b16a1a7f6781e4f65f763e6cf35ca8c4382a4b/research/object_detection/models/ssd_resnet_v1_ppn_feature_extractor.py#L113-L114

In order to show these images properly in tensorboard, the values should be in the range [0, 1]. The image is added to tensorboard here:
https://github.com/tensorflow/models/blob/52b16a1a7f6781e4f65f763e6cf35ca8c4382a4b/research/object_detection/model_lib_v2.py#L273-L277

It would be possible to undo this scaling so that the images are in the value range [0, 1], so that they are correctly represented.
Or maybe we could represent the original image before preprocessing, simply divided by 255.

**Comment by turowicz** [comment]
@tombstone @jch1 @pkulzc any ideas?

**Comment by turowicz** [comment]
@carlos-vl do you know if this kind of images end up in the training or is it just to show them in tensorboard?

**Comment by pascal-hwky** [comment]
I'm getting the same issues (using `ssd_mobilenet_v2_keras` feature extractor that applies -1 to 1 normalization) during training. The evaluation images are visualized correctly. @turowicz these artifacts should not be a problem for your training, it only affects visualization as `tf.summary.image` expects images to be normalized between 0 and 1.

**Comment by turowicz** [comment]
OK thanks

**Comment by DevLob-zz** [comment]
> I think I found the reason for this behaviour. The input data is encoded as 8bit jpeg or png in the tfrecord files (at least in my case). The images are scaled during the preprocessing step of the feature extractors. Each feature extractor has a different scaling apparently. For example:
> 
> https://github.com/tensorflow/models/blob/52b16a1a7f6781e4f65f763e6cf35ca8c4382a4b/research/object_detection/models/ssd_efficientnet_bifpn_feature_extractor.py#L200-L202
> 
> https://github.com/tensorflow/models/blob/52b16a1a7f6781e4f65f763e6cf35ca8c4382a4b/research/object_detection/models/ssd_resnet_v1_ppn_feature_extractor.py#L113-L114
> 
> In order to show these images properly in tensorboard, the values should be in the range [0, 1]. The image is added to tensorboard here:
> https://github.com/tensorflow/models/blob/52b16a1a7f6781e4f65f763e6cf35ca8c4382a4b/research/object_detection/model_lib_v2.py#L273-L277
> 
> It would be possible to undo this scaling so that the images are in the value range [0, 1], so that they are correctly represented.
> Or maybe we could represent the original image before preprocessing, simply divided by 255.

This great Many thanks

**Comment by Meriipu** [comment]
This is still an issue so probably should not be closed.
